### PR TITLE
[doc] Add shape about layout improvements

### PR DESCRIPTION
**Pull request by gcoutable**
Bug:  #1923 
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/iterations/2023.6/scopes/improveEmptyNodesLayout.adoc [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+- [ ] Shape written, reviewed and merged
+- [ ] ADR written, reviewed and merged
+- [ ] Mock of the features done with Figma
+- [ ] First prototype of the user interface
+- [ ] First version of the user interface done with Cypress tests
+
+== Second task
+
+- [ ] Capture proper screenshots for the CHANGELOG and the documentation
+- [ ] User guide updated with screenshots
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/scopes/improveEmptyNodesLayout.adoc]: No need for a scope right know (especially given that it does not contain anything relevant for the iteration)

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+- ELK: Thanks to configuration, it computes layout for diagrams.
+In sirius-components, we are using it to compute the first layout of a diagram, or to perfrom an "arrange all" on demand.
+- Incremental Layout: Our layout algorithm used to handle layout updates of diagrams during their life.
+
+Having two layout solutions bring some issues.
+One of them is about ELK. Sometime, ELK needs to be twisted to get the result we want.
+An other one concerns the moment we goes from ELK layout to the incremental layout. The incremental layout needs to update only the part that needs to be updated without touching anything else ELK produced.
+
+Lately, we were not able to configure ELK to let it produce a fine layout without spending a lot of time to implements a lot of workarounds to twist ELK and make it produced the layout we want.
+We do not want to use ELK anymore. For that we need to be able to handle the lifecycle of a diagram layout from its creation to its desctruction.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: One sentence per line

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
+= (L) Shape
+
+== Problem
+
+It has been some iterations we have increase the effort put on diagram layout.
+We try as much as we can to maintain two layout solutions:
+
+- ELK: Thanks to configuration, it computes layout for diagrams.
+In sirius-components, we are using it to compute the first layout of a diagram, or to perfrom an "arrange all" on demand.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: perform

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+- Incremental Layout: Our layout algorithm used to handle layout updates of diagrams during their life.
+
+Having two layout solutions bring some issues.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: brings

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+One of them is about ELK. Sometime, ELK needs to be twisted to get the result we want.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: One sentence per line

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+An other one concerns the moment we goes from ELK layout to the incremental layout. The incremental layout needs to update only the part that needs to be updated without touching anything else ELK produced.
+
+Lately, we were not able to configure ELK to let it produce a fine layout without spending a lot of time to implements a lot of workarounds to twist ELK and make it produced the layout we want.
+We do not want to use ELK anymore. For that we need to be able to handle the lifecycle of a diagram layout from its creation to its desctruction.
+
+== Key Result
+
+Empty nodes can be displayed without overlap.
+The nodes should be able to have various descendants.
+Nodes size and position should take into account, data comming from the specifier, updates comming from the user like the move or resize and, default values.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: coming

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+The user wants to create a semantic element from the explorer.
+The user uses the object creation modal on an explorer element.
+The node will be display at abscissa 0 but under the farthest most bottom node.
+
+[NOTE]
+====
+
+Using a tool that creates many nodes will use the conjuction of the scenario <<scenario-1,1>> and <<scenario-4,4>> above.
+If a tool creates many nodes on the same deep, one of the node will be located where the user clicked, others node will be display in column (or row) under (like scenario 1)
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: at the same depth

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+1- Produce a layout with default size for node
+2- Move event
+3- Resize event
+4- Single click event
+5- Extract the calculation of height and width expressions out of the renderer, to use data from view/odesign
+
+== Rabbit holes
+
+We should keep in mind that a tool can do anything, create, update or, remove elements. It is not because a tool is called a node creation tool that it will only be used to create something (it may be should but...)
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: One sentence per line

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+Using a tool that creates many nodes will use the conjuction of the <<scenario-1,scenario 1>> and <<scenario-4,scenario 4>>.
+If a tool creates many nodes on the same deep, one of the node will be located where the user clicked, others node will be display following the direction (like <<scenario-1,scenario 1>>)
+If a tool creates many nodes, a node and one or many children, the new parent node will be created where the user clicked, children will be display starting at (0, 0) in the parent (like <<scenario-4,scenario 4>>)
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Copy / Paste

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+Shrinking a node by moving North or West anchor can be tricky because it changes the size of the node _and_ it's position.
```
Comment (sbegaudeau) [inline doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: moving its

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 3 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 4 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 5 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 2 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 6 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 7 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Done

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 8 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Done

File: doc/iterations/2023.6/scopes/improveEmptyNodesLayout.adoc (hunk 1 above)
Comment (gcoutable) [reply doc/iterations/2023.6/scopes/improveEmptyNodesLayout.adoc]: Removed

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 9 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: Note in scenario 4 removed

File: doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc (hunk 10 above)
Comment (gcoutable) [reply doc/iterations/2023.6/shapes/improveEmptyNodesLayout.adoc]: done

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**